### PR TITLE
Support rpi3 and rpi 4 boards

### PR DESCRIPTION
**Pull request by Itxaka**
Due to rpi3 not supporting GPT disks, we need to separate the building into rpi3 and 4 so rpi4 gets GPT disks and can auto-expand the last partition properly (and add extra partitions and so on)

This also includes some cleanup for the luet cache dirs, includes the kairos-agent which was missing from the image for pulling images, and cleans up the lvm vgs if they were left

Fixes https://github.com/kairos-io/kairos/issues/1636

### CODE REVIEW

**Comment by Itxaka** [comment]
rpi3 image output:

```
Disk output/rpi3.raw: 7,26 GiB, 7790592000 bytes, 15216000 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: dos
Disk identifier: 0x00000000

Old situation:

Device            Boot    Start      End  Sectors  Size Id Type
output/image.raw1          8192   204799   196608   96M  c W95 FAT32 (LBA)
output/image.raw2        204800 10428415 10223616  4,9G 83 Linux
output/image.raw3      10428416 15048703  4620288  2,2G 8e Linux LVM
output/image.raw4      15048704 15179775   131072   64M 83 Linux
```

rpi4 image output:

```
Disk output/rpi4.raw: 7,26 GiB, 7790592000 bytes, 15216000 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: gpt
Disk identifier: DC92B7A1-7769-2840-A7C4-6434F08974A3

Old situation:

Device              Start      End  Sectors  Size Type
output/rpi4.raw1     8192   204799   196608   96M Microsoft basic data
output/rpi4.raw2   204800 10428415 10223616  4,9G Linux filesystem
output/rpi4.raw3 10428416 14917631  4489216  2,1G Linux filesystem
output/rpi4.raw4 14917632 15048703   131072   64M Linux filesystem
output/rpi4.raw5 15048704 15179775   131072   64M Linux filesystem

```

**Comment by Itxaka** [comment]
going to test this with rpi4 but I dont have access to an rpi3 for testing that. @jimmykarily or @mauromorales do you fine folk have an rpi3 that can test the images built with this?

File: tools-image/build-arm-image.sh [hunk 1]
```diff
@@ -473,7 +494,11 @@ sync
 sleep 5
 sync
 
-kpartx -dv $DRIVE || true
+if [ "$model" == 'rpi4' ]; then
+  kpartx -dvg $DRIVE
```
Comment (mauromorales) [inline tools-image/build-arm-image.sh:498]:
```suggestion
  kpartx -dvg $DRIVE || true
```

maybe?

**Review by mauromorales — APPROVED** [review]
lgtm

File: tools-image/build-arm-image.sh (hunk 1 above)
Comment (mauromorales) [reply tools-image/build-arm-image.sh:498]: oki then no need

Comment (Itxaka) [reply tools-image/build-arm-image.sh:498]: I dunno, tdo you think its necessary? Does this usually fail?

I want to clean this up in order to be able to runs it several times locally with no failures and it seems to work as expected, but of course if it fails somewhere else...

**Comment by Itxaka** [comment]
on rpi4, built with this osbuilder branch and some kairos changes to fix the model to rpi4 (`--MODEL=rpi4`), the system boots as expected and the partitions are there as expected, immucore loads everything, mounts everything, stages run, last partition (persistent) is fully expanded on first boot:


```
kairos@localhost:~> sudo lsblk
NAME  MAJ:MIN RM  SIZE RO TYPE MOUNTPOINTS
loop0   7:0    0    2G  1 loop /
sda     8:0    1 58.6G  0 disk 
├─sda1
│       8:1    1   96M  0 part 
├─sda2
│       8:2    1  6.1G  0 part /run/initramfs/cos-state
├─sda3
│       8:3    1  4.1G  0 part 
├─sda4
│       8:4    1   64M  0 part /oem
└─sda5
        8:5    1 48.3G  0 part /usr/local/.state/usr-share-pki-trust.bind/anchors
                               /usr/share/pki/trust/anchors
                               /usr/share/pki/trust
                               /var/lib/wicked
                               /var/lib/snapd
                               /var/lib/rancher
                               /var/lib/longhorn
                               /var/lib/kubelet
                               /var/lib/dbus
                               /var/lib/cni
                               /var/lib/ca-certificates
                               /etc/zfs
                               /etc/systemd
                               /etc/sysconfig
                               /etc/ssh
                               /var/snap
                               /etc/runlevels
                               /etc/rancher
                               /etc/modprobe.d
                               /var/log
                               /usr/libexec
                               /etc/kubernetes
                               /etc/iscsi
                               /etc/cni
                               /root
                               /opt
                               /home
                               /usr/local
```

```
time="2023-04-06T12:00:14Z" level=info msg="Processing stage step 'Grow persistent'. ( commands: 0, files: 0, ... )"
time="2023-04-06T12:00:16Z" level=info msg="Extending last partition to max space"
time="2023-04-06T12:00:17Z" level=info msg="Done executing stage 'rootfs.after'\n"
```

**Comment by Itxaka** [comment]
We just need someone to build this for rpi3 and confim that it works as expected :)

steps to reproduce.

 - clone this branch locally
 - build the osbuilder and push it to ttl.sh
```
$ docker build -t ttl.sh/osbuilder:rpi3test .
$ docker push ttl.sh/osbuilder:rpi3test
```
 - Apply this diff to the kairos earthfile:
```
diff --git a/Earthfile b/Earthfile
index d0fbad9..6764c44 100644
--- a/Earthfile
+++ b/Earthfile
@@ -27,7 +27,7 @@ ARG COSIGN_EXPERIMENTAL=0
 ARG CGO_ENABLED=0
 # renovate: datasource=docker depName=quay.io/kairos/osbuilder-tools versioning=semver-coerced
 ARG OSBUILDER_VERSION=v0.8.2
-ARG OSBUILDER_IMAGE=quay.io/kairos/osbuilder-tools:$OSBUILDER_VERSION
+ARG OSBUILDER_IMAGE=ttl.sh/osbuilder:itxaka
 ARG GOLINT_VERSION=1.52.2
 # renovate: datasource=docker depName=golang
 ARG GO_VERSION=1.20
@@ -66,17 +66,17 @@ ci:
 
 all-arm:
   ARG SECURITY_SCANS=true
-  BUILD --platform=linux/arm64 +base-image --MODEL=rpi64
+  BUILD --platform=linux/arm64 +base-image --MODEL=rpi4
   IF [ "$SECURITY_SCANS" = "true" ]
-      BUILD --platform=linux/arm64 +image-sbom --MODEL=rpi64
-      BUILD --platform=linux/arm64 +trivy-scan --MODEL=rpi64
-      BUILD --platform=linux/arm64 +grype-scan --MODEL=rpi64
+      BUILD --platform=linux/arm64 +image-sbom --MODEL=rpi4
+      BUILD --platform=linux/arm64 +trivy-scan --MODEL=rpi4
+      BUILD --platform=linux/arm64 +grype-scan --MODEL=rpi4
   END
   
   IF [[ "$FLAVOR" = "ubuntu-20-lts-arm-nvidia-jetson-agx-orin" ]]
-    BUILD +prepare-arm-image --MODEL=rpi64 --FLAVOR=${FLAVOR}
+    BUILD +prepare-arm-image --MODEL=rpi4 --FLAVOR=${FLAVOR}
   ELSE
-    BUILD +arm-image --MODEL=rpi64
+    BUILD +arm-image --MODEL=rpi4
   END
 
 arm-container-image:
@@ -239,7 +239,7 @@ framework:
     COPY overlay/files /framework
 
     # Copy common overlay files for Raspberry Pi
-    IF [ "$MODEL" = "rpi64" ]
+    IF [ "$MODEL" = "rpi3" ] || [ "$MODEL" = "rpi4" ]
         COPY overlay/files-rpi/ /framework
     END
 
@@ -632,7 +632,7 @@ arm-image:
   ARG COMPRESS_IMG=true
   ARG IMG_COMPRESSION=xz
   FROM $OSBUILDER_IMAGE
-  ARG MODEL=rpi64
+  ARG MODEL=rpi4
   COPY +version/VERSION ./
   RUN echo "version ${VERSION}"
   ARG VERSION=$(cat VERSION)
@@ -675,7 +675,7 @@ prepare-arm-image:
   ARG OSBUILDER_IMAGE
   ARG COMPRESS_IMG=true
   FROM $OSBUILDER_IMAGE
-  ARG MODEL=rpi64
+  ARG MODEL=rpi4
   ARG IMAGE_NAME=${FLAVOR}.img
   WORKDIR /build
   # These sizes are in MB

```
 - Build the arm image with the model set to rpi3:
`earthly -P +all-arm --MODEL=rpi3 --SECURITY_SCANS=false --FLAVOR=opensuse-leap-arm-rpi --COMPRESS_IMG=false`
 - Burn the resulting image onto an sd card and test that it boots and everything is mounted in the rigth places, partitions are correct, labels are there.
 - ???????
 - PROFIT!

**Comment by jimmykarily** [comment]
I don't have an rpi3 around.

**Comment by mauromorales** [comment]
Ludea mentioned on Slack, that he will try to find time to give it a shot

**Comment by mudler** [comment]
> We just need someone to build this for rpi3 and confim that it works as expected :)
> 
> steps to reproduce.
> 
>     * clone this branch locally
> 
>     * build the osbuilder and push it to ttl.sh
> 
> 
> ```
> $ docker build -t ttl.sh/osbuilder:rpi3test .
> $ docker push ttl.sh/osbuilder:rpi3test
> ```
> 
>     * Apply this diff to the kairos earthfile:
> 
> 
> ```
> diff --git a/Earthfile b/Earthfile
> index d0fbad9..6764c44 100644
> --- a/Earthfile
> +++ b/Earthfile
> @@ -27,7 +27,7 @@ ARG COSIGN_EXPERIMENTAL=0
>  ARG CGO_ENABLED=0
>  # renovate: datasource=docker depName=quay.io/kairos/osbuilder-tools versioning=semver-coerced
>  ARG OSBUILDER_VERSION=v0.8.2
> -ARG OSBUILDER_IMAGE=quay.io/kairos/osbuilder-tools:$OSBUILDER_VERSION
> +ARG OSBUILDER_IMAGE=ttl.sh/osbuilder:itxaka
>  ARG GOLINT_VERSION=1.52.2
>  # renovate: datasource=docker depName=golang
>  ARG GO_VERSION=1.20
> @@ -66,17 +66,17 @@ ci:
>  
>  all-arm:
>    ARG SECURITY_SCANS=true
> -  BUILD --platform=linux/arm64 +base-image --MODEL=rpi64
> +  BUILD --platform=linux/arm64 +base-image --MODEL=rpi4
>    IF [ "$SECURITY_SCANS" = "true" ]
> -      BUILD --platform=linux/arm64 +image-sbom --MODEL=rpi64
> -      BUILD --platform=linux/arm64 +trivy-scan --MODEL=rpi64
> -      BUILD --platform=linux/arm64 +grype-scan --MODEL=rpi64
> +      BUILD --platform=linux/arm64 +image-sbom --MODEL=rpi4
> +      BUILD --platform=linux/arm64 +trivy-scan --MODEL=rpi4
> +      BUILD --platform=linux/arm64 +grype-scan --MODEL=rpi4
>    END
>    
>    IF [[ "$FLAVOR" = "ubuntu-20-lts-arm-nvidia-jetson-agx-orin" ]]
> -    BUILD +prepare-arm-image --MODEL=rpi64 --FLAVOR=${FLAVOR}
> +    BUILD +prepare-arm-image --MODEL=rpi4 --FLAVOR=${FLAVOR}
>    ELSE
> -    BUILD +arm-image --MODEL=rpi64
> +    BUILD +arm-image --MODEL=rpi4
>    END
>  
>  arm-container-image:
> @@ -239,7 +239,7 @@ framework:
>      COPY overlay/files /framework
>  
>      # Copy common overlay files for Raspberry Pi
> -    IF [ "$MODEL" = "rpi64" ]
> +    IF [ "$MODEL" = "rpi3" ] || [ "$MODEL" = "rpi4" ]
>          COPY overlay/files-rpi/ /framework
>      END
>  
> @@ -632,7 +632,7 @@ arm-image:
>    ARG COMPRESS_IMG=true
>    ARG IMG_COMPRESSION=xz
>    FROM $OSBUILDER_IMAGE
> -  ARG MODEL=rpi64
> +  ARG MODEL=rpi4
>    COPY +version/VERSION ./
>    RUN echo "version ${VERSION}"
>    ARG VERSION=$(cat VERSION)
> @@ -675,7 +675,7 @@ prepare-arm-image:
>    ARG OSBUILDER_IMAGE
>    ARG COMPRESS_IMG=true
>    FROM $OSBUILDER_IMAGE
> -  ARG MODEL=rpi64
> +  ARG MODEL=rpi4
>    ARG IMAGE_NAME=${FLAVOR}.img
>    WORKDIR /build
>    # These sizes are in MB
> ```
> 
>     * Build the arm image with the model set to rpi3:
>       `earthly -P +all-arm --MODEL=rpi3 --SECURITY_SCANS=false --FLAVOR=opensuse-leap-arm-rpi --COMPRESS_IMG=false`
> 
>     * Burn the resulting image onto an sd card and test that it boots and everything is mounted in the rigth places, partitions are correct, labels are there.
> 
>     * ???????
> 
>     * PROFIT!

I have an RPI3 that I can use for tests - but way better if you can provide the flashable artifact directly (maybe as an OCI image?)

**Comment by mudler** [comment]
@Itxaka just tested the image on RPI3 and boots fine !

**Review by mudler — APPROVED** [review]

**Comment by Itxaka** [comment]
Thanks for testing @mudler !

**Pull request merged**